### ISSUE TITLE
Parsing optimization or lazy loading

### ISSUE THREAD

**Issue by Tihon-Ustinov** [Kotlin/kotlinx.serialization#1432]
I have a serious performance issue when parsing.

To demonstrate how it works, I wrote the following code to compare the performance of native `JSON.parse` and `Json { ignoreUnknownKeys = true }.decodeFromString`

![image](https://user-images.githubusercontent.com/15249713/115868243-d3053f80-a444-11eb-9b11-29cfae254ac2.png)

After that, I perform a time measurement via chrome profile and get this result

`Json { ignoreUnknownKeys = true }.decodeFromString` 1.92s
![image](https://user-images.githubusercontent.com/15249713/115868671-79e9db80-a445-11eb-85bb-30fe0fd63bf5.png)

`JSON.parse` 23ms
![image](https://user-images.githubusercontent.com/15249713/115868789-a6055c80-a445-11eb-8d6a-028da27f7455.png)

Is it possible to somehow speed up the work? Use partial JSON.parse in js implementation?
For example, create a Kotlin DTO class and populate it like this?

```kotlin
@Serializable
class Foo {
  val bas: String
}
```

```js
let parseData = JSON.parse ('{"bas": "hello"}')
Object.assing (new Foo (), parseData)
```

is it possible to somehow speed it up?

**Comment by elizarov** [comment]
Some way to get better speed in Kotlin/JS serialization:
1. Don't create `Json` instance every time. Cache it with `val json = JSON { ... my options ... }` 
2. Use `JSON.parse` followed by `json.decodeFromDynamic` (see https://kotlin.github.io/kotlinx.serialization/kotlinx-serialization-json/kotlinx-serialization-json/kotlinx.serialization.json/decode-from-dynamic.html)

**Comment by Tihon-Ustinov** [comment]
the option you proposed takes even longer

![image](https://user-images.githubusercontent.com/15249713/115875032-7bb79d00-a44d-11eb-80c4-8d1a2b13779c.png)

![image](https://user-images.githubusercontent.com/15249713/115875086-85d99b80-a44d-11eb-867d-0aef2610ce6a.png)

**Comment by elizarov** [comment]
But what is the JSON object? Can you provide a self-contained example that we can use to repeat yours measurements?

**Comment by Tihon-Ustinov** [comment]
More json https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.js/-j-s-o-n/

Unfortunately, I cannot provide you with this particular JSON as it is part of the NDA.

But I can show pieces that could slow down the parsing

![image](https://user-images.githubusercontent.com/15249713/115879987-e8816600-a452-11eb-9cbf-cf007ababa5f.png)

**Comment by EmmanueleVilla** [comment]
This is still relevant, deserializing with kotlinx.serialization is 20x slower than using JSON.parse
This is an old example with a 200kb json: https://github.com/SerVB/kx-serialization-js-benchmark
but the timings are the same with the latest libraries

**Comment by ApoloApps** [comment]
@sandwwraith any progress on this? are there any short term plans for this? Thanks

**Comment by artemBondar3SS** [comment]
Is there any estimation for this to be fixed/improved?

**Comment by sandwwraith** [comment]
Sadly, there are no immediate plans for this. I'd also like to point out that most Kotlin objects contain additional information, so we need to create them during deserialization — I suppose that's where the main difference is coming from.

I suggest trying [decodeFromDynamic](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/decode-from-dynamic.html) over the result of `JSON.parse`, maybe this will be faster in your case. Just remember to check the results, because they may slightly differ.

**Comment by ApoloApps** [comment]
>I suggest trying [decodeFromDynamic](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/decode-from-dynamic.html)

cant find it for wasmjs

**Comment by sandwwraith** [comment]
@ApoloApps Indeed, it was not adopted for wasm. Can you open a new issue please?
